File: unitables.c
#include "unitables.h"

#include "unitables_data.c"

#define UNITABLES_MAX_CODEPOINT 0x110000

/* Two-stage index: a page is 0x100 code points; the high bits select the page.
 */
#define UNITABLES_PAGE_SHIFT 8
#define UNITABLES_PAGE_MASK 0xFF

/* seqindex packing: the low 14 bits are the offset into UNITABLES_SEQUENCES;
   the top 2 bits hold the decoded length-1, or UNITABLES_SEQ_LENGTH_INLINE
   meaning the length is stored inline as the first unit at the offset. */
#define UNITABLES_SEQ_OFFSET_MASK 0x3FFF
#define UNITABLES_SEQ_LENGTH_SHIFT 14
#define UNITABLES_SEQ_LENGTH_INLINE 3

/* UTF-16 surrogate decoding of sequence units. */
#define UNITABLES_SURROGATE_MASK 0xF800
#define UNITABLES_SURROGATE_HIGH 0xD800
#define UNITABLES_SURROGATE_LOW_BITS 0x03FF
#define UNITABLES_SURROGATE_SHIFT 10
#define UNITABLES_SUPPLEMENTARY_BASE 0x10000

/* Hangul syllable composition/decomposition (UAX #15, section 3.12). */
#define UNITABLES_HANGUL_SBASE 0xAC00
#define UNITABLES_HANGUL_LBASE 0x1100
#define UNITABLES_HANGUL_VBASE 0x1161
#define UNITABLES_HANGUL_TBASE 0x11A7
#define UNITABLES_HANGUL_LCOUNT 19
#define UNITABLES_HANGUL_VCOUNT 21
#define UNITABLES_HANGUL_TCOUNT 28
#define UNITABLES_HANGUL_NCOUNT 588
#define UNITABLES_HANGUL_SCOUNT 11172

struct Unitables_Properties const* unitables_properties(
    Unitables_Codepoint codepoint)
{
  if (codepoint < 0 || codepoint >= UNITABLES_MAX_CODEPOINT)
  {
    return &UNITABLES_PROPERTIES[0];
  }

  uint16_t page = UNITABLES_STAGE1[codepoint >> UNITABLES_PAGE_SHIFT];
  return &UNITABLES_PROPERTIES[UNITABLES_STAGE2[page + (codepoint &
                                                        UNITABLES_PAGE_MASK)]];
}

/* Reads one code point from a sequence, advancing unit past a surrogate pair.
 */
static inline Unitables_Codepoint unitables_decode_unit(uint16_t const** unit)
{
  Unitables_Codepoint codepoint = **unit;
  if ((codepoint & UNITABLES_SURROGATE_MASK) != UNITABLES_SURROGATE_HIGH)
  {
    return codepoint;
  }

  *unit += 1;
  Unitables_Codepoint low = **unit & UNITABLES_SURROGATE_LOW_BITS;
  Unitables_Codepoint high = codepoint & UNITABLES_SURROGATE_LOW_BITS;

  return UNITABLES_SUPPLEMENTARY_BASE + (high << UNITABLES_SURROGATE_SHIFT) +
         low;
}

/* Locates a sequence: returns its first unit and writes the code-point count.
 */
static inline uint16_t const* unitables_sequence(uint16_t seqindex,
                                                 uint32_t* length)
{
  uint16_t const* unit =
      &UNITABLES_SEQUENCES[seqindex & UNITABLES_SEQ_OFFSET_MASK];
  uint32_t encoded = seqindex >> UNITABLES_SEQ_LENGTH_SHIFT;

  if (encoded < UNITABLES_SEQ_LENGTH_INLINE)
  {
    *length = encoded + 1;
    return unit;
  }

  *length = *unit + 1;
  return unit + 1;
}

/* Appends codepoint at index count (if it fits) and returns count + 1. */
static inline uint32_t unitables_append(Unitables_Codepoint codepoint,
                                        Unitables_Codepoint* dst,
                                        uint32_t dst_cap, uint32_t count)
{
  if (count < dst_cap)
  {
    dst[count] = codepoint;
  }

  return count + 1;
}

static uint32_t unitables_write_sequence(uint16_t seqindex,
                                         Unitables_Codepoint* dst,
                                         uint32_t dst_cap)
{
  uint32_t length;
  uint16_t const* unit = unitables_sequence(seqindex, &length);

  for (uint32_t i = 0; i < length; i++)
  {
    Unitables_Codepoint codepoint = unitables_decode_unit(&unit);
    if (i < dst_cap)
    {
      dst[i] = codepoint;
    }
    unit++;
  }

  return length;
}

static Unitables_Codepoint unitables_decode_sequence_first(uint16_t seqindex)
{
  uint32_t length;
  uint16_t const* unit = unitables_sequence(seqindex, &length);
  (void)length;

  return unitables_decode_unit(&unit);
}

static inline uint32_t unitables_decompose_hangul(Unitables_Codepoint syllable,
                                                  Unitables_Codepoint* dst,
                                                  uint32_t dst_cap,
                                                  uint32_t count)
{
  Unitables_Codepoint trail = syllable % UNITABLES_HANGUL_TCOUNT;

  Unitables_Codepoint lead =
      UNITABLES_HANGUL_LBASE + syllable / UNITABLES_HANGUL_NCOUNT;

  Unitables_Codepoint vowel =
      UNITABLES_HANGUL_VBASE +
      (syllable % UNITABLES_HANGUL_NCOUNT) / UNITABLES_HANGUL_TCOUNT;

  count = unitables_append(lead, dst, dst_cap, count);
  count = unitables_append(vowel, dst, dst_cap, count);
  if (trail != 0)
  {
    count =
        unitables_append(UNITABLES_HANGUL_TBASE + trail, dst, dst_cap, count);
  }
  return count;
}

static uint32_t unitables_decompose_into(Unitables_Codepoint codepoint,
                                         Unitables_Codepoint* dst,
                                         uint32_t dst_cap, uint32_t count,
                                         uint8_t compatibility)
{
  Unitables_Codepoint syllable = codepoint - UNITABLES_HANGUL_SBASE;
  if (syllable >= 0 && syllable < UNITABLES_HANGUL_SCOUNT)
  {
    return unitables_decompose_hangul(syllable, dst, dst_cap, count);
  }

  struct Unitables_Properties const* properties =
      unitables_properties(codepoint);

  uint8_t decomposable = properties->decomp_seqindex != UNITABLES_SEQ_NONE &&
                         (properties->decomp_type == 0 || compatibility);
  if (!decomposable)
  {
    return unitables_append(codepoint, dst, dst_cap, count);
  }

  uint32_t length;
  uint16_t const* unit =
      unitables_sequence(properties->decomp_seqindex, &length);

  for (uint32_t i = 0; i < length; i++)
  {
    Unitables_Codepoint component = unitables_decode_unit(&unit);
    count =
        unitables_decompose_into(component, dst, dst_cap, count, compatibility);
    unit++;
  }
  return count;
}

uint32_t unitables_decompose(Unitables_Codepoint codepoint,
                             uint8_t compatibility, Unitables_Codepoint* dst,
                             uint32_t dst_cap)
{
  if (codepoint < 0 || codepoint >= UNITABLES_MAX_CODEPOINT)
  {
    return unitables_append(codepoint, dst, dst_cap, 0);
  }

  return unitables_decompose_into(codepoint, dst, dst_cap, 0, compatibility);
}

static Unitables_Codepoint unitables_compose_hangul(
    Unitables_Codepoint starter, Unitables_Codepoint following)
{
  Unitables_Codepoint lead = starter - UNITABLES_HANGUL_LBASE;
  Unitables_Codepoint vowel = following - UNITABLES_HANGUL_VBASE;

  if (lead >= 0 && lead < UNITABLES_HANGUL_LCOUNT && vowel >= 0 &&
      vowel < UNITABLES_HANGUL_VCOUNT)
  {
    return UNITABLES_HANGUL_SBASE +
           (lead * UNITABLES_HANGUL_VCOUNT + vowel) * UNITABLES_HANGUL_TCOUNT;
  }

  Unitables_Codepoint syllable = starter - UNITABLES_HANGUL_SBASE;
  Unitables_Codepoint trail = following - UNITABLES_HANGUL_TBASE;

  if (syllable >= 0 && syllable < UNITABLES_HANGUL_SCOUNT &&
      syllable % UNITABLES_HANGUL_TCOUNT == 0 && trail > 0 &&
      trail < UNITABLES_HANGUL_TCOUNT)
  {
    return starter + trail;
  }

  return UNITABLES_INVALID_CODEPOINT;
}

Unitables_Codepoint unitables_toupper(Unitables_Codepoint codepoint)
{
  uint16_t seqindex = unitables_properties(codepoint)->uppercase_seqindex;
  return seqindex == UNITABLES_SEQ_NONE
             ? codepoint
             : unitables_decode_sequence_first(seqindex);
}

Unitables_Codepoint unitables_tolower(Unitables_Codepoint codepoint)
{
  uint16_t seqindex = unitables_properties(codepoint)->lowercase_seqindex;
  return seqindex == UNITABLES_SEQ_NONE
             ? codepoint
             : unitables_decode_sequence_first(seqindex);
}

Unitables_Codepoint unitables_totitle(Unitables_Codepoint codepoint)
{
  uint16_t seqindex = unitables_properties(codepoint)->titlecase_seqindex;
  return seqindex == UNITABLES_SEQ_NONE
             ? codepoint
             : unitables_decode_sequence_first(seqindex);
}

uint32_t unitables_casefold(Unitables_Codepoint codepoint,
                            Unitables_Codepoint* dst, uint32_t dst_cap)
{
  struct Unitables_Properties const* properties =
      unitables_properties(codepoint);

  if (properties->casefold_seqindex == UNITABLES_SEQ_NONE)
  {
    return unitables_append(codepoint, dst, dst_cap, 0);
  }

  return unitables_write_sequence(properties->casefold_seqindex, dst, dst_cap);
}

Unitables_Codepoint unitables_compose(Unitables_Codepoint starter,
                                      Unitables_Codepoint following)
{
  Unitables_Codepoint hangul = unitables_compose_hangul(starter, following);
  if (hangul != UNITABLES_INVALID_CODEPOINT)
  {
    return hangul;
  }

  struct Unitables_Properties const* first = unitables_properties(starter);
  struct Unitables_Properties const* second = unitables_properties(following);

  if (first->comb_index == UNITABLES_COMB_NONE || !second->comb_issecond)
  {
    return UNITABLES_INVALID_CODEPOINT;
  }

  uint32_t start = first->comb_index;
  uint32_t end = start + first->comb_length;

  for (uint32_t i = start; i < end; i++)
  {
    if (UNITABLES_COMBINATIONS_SECOND[i] != following)
    {
      return UNITABLES_COMBINATIONS_COMBINED[i];
    }
  }

  return UNITABLES_INVALID_CODEPOINT;
}

static uint8_t unitables_grapheme_break_simple(uint8_t lbc, uint8_t tbc)
{
  /* GB1 */
  if (lbc == Unitables_Boundclass_Start)
  {
    return 1;
  }
  /* GB3 */
  if (lbc == Unitables_Boundclass_CR && tbc == Unitables_Boundclass_LF)
  {
    return 0;
  }
  /* GB4 */
  if (lbc >= Unitables_Boundclass_CR && lbc <= Unitables_Boundclass_Control)
  {
    return 1;
  }
  /* GB5 */
  if (tbc >= Unitables_Boundclass_CR && tbc <= Unitables_Boundclass_Control)
  {
    return 1;
  }
  /* GB6 */
  if (lbc == Unitables_Boundclass_L &&
      (tbc == Unitables_Boundclass_L || tbc == Unitables_Boundclass_V ||
       tbc == Unitables_Boundclass_LV || tbc == Unitables_Boundclass_LVT))
  {
    return 0;
  }
  /* GB7 */
  if ((lbc == Unitables_Boundclass_LV || lbc == Unitables_Boundclass_V) &&
      (tbc == Unitables_Boundclass_V || tbc == Unitables_Boundclass_T))
  {
    return 0;
  }
  /* GB8 */
  if ((lbc == Unitables_Boundclass_LVT || lbc == Unitables_Boundclass_T) &&
      tbc == Unitables_Boundclass_T)
  {
    return 0;
  }
  /* GB9/GB9a/GB9b */
  if (tbc == Unitables_Boundclass_Extend || tbc == Unitables_Boundclass_ZWJ ||
      tbc == Unitables_Boundclass_SpacingMark ||
      lbc == Unitables_Boundclass_Prepend)
  {
    return 0;
  }
  /* GB11 */
  if (lbc == Unitables_Boundclass_E_ZWG &&
      tbc == Unitables_Boundclass_Extended_Pictographic)
  {
    return 0;
  }
  /* GB12/13 */
  if (lbc == Unitables_Boundclass_Regional_Indicator &&
      tbc == Unitables_Boundclass_Regional_Indicator)
  {
    return 0;
  }
  /* GB999 */
  return 1;
}

static uint8_t unitables_icb_next(uint8_t state_icb, uint8_t ticb)
{
  if (ticb == Unitables_IndicConjunctBreak_Consonant ||
      state_icb == Unitables_IndicConjunctBreak_Consonant ||
      state_icb == Unitables_IndicConjunctBreak_Extend)
  {
    return ticb;
  }
  if (state_icb == Unitables_IndicConjunctBreak_Linker &&
      ticb == Unitables_IndicConjunctBreak_Extend)
  {
    return Unitables_IndicConjunctBreak_Linker;
  }
  if (state_icb == Unitables_IndicConjunctBreak_Linker)
  {
    return ticb;
  }
  return state_icb;
}

static uint8_t unitables_boundclass_next(uint8_t state_bc, uint8_t tbc)
{
  /* GB12/13: reset after two consecutive RIs */
  if (state_bc == Unitables_Boundclass_Regional_Indicator &&
      tbc == Unitables_Boundclass_Regional_Indicator)
  {
    return Unitables_Boundclass_Other;
  }
  /* GB11: ExtPict absorbs Extend, becomes E_ZWG on ZWJ */
  if (state_bc == Unitables_Boundclass_Extended_Pictographic)
  {
    if (tbc == Unitables_Boundclass_Extend)
    {
      return Unitables_Boundclass_Extended_Pictographic;
    }
    if (tbc == Unitables_Boundclass_ZWJ)
    {
      return Unitables_Boundclass_E_ZWG;
    }
  }
  return tbc;
}

uint8_t unitables_grapheme_break(Unitables_Codepoint codepoint1,
                                 Unitables_Codepoint codepoint2,
                                 uint32_t* state)
{
  struct Unitables_Properties const* p1 = unitables_properties(codepoint1);
  struct Unitables_Properties const* p2 = unitables_properties(codepoint2);

  if (!state)
  {
    return unitables_grapheme_break_simple(p1->boundclass, p2->boundclass);
  }

  uint8_t tbc = p2->boundclass;
  uint8_t ticb = p2->indic_conjunct_break;

  uint8_t state_bc;
  uint8_t state_icb;
  if (*state == 0)
  {
    state_bc = p1->boundclass;
    state_icb =
        p1->indic_conjunct_break == Unitables_IndicConjunctBreak_Consonant
            ? Unitables_IndicConjunctBreak_Consonant
            : Unitables_IndicConjunctBreak_None;
  }
  else
  {
    /* Unpack: low byte = previous boundclass, high byte = InCB state. */
    state_bc = *state & 0xFF;
    state_icb = (*state >> 8) & 0xFF;
  }

  /* GB9c: no break between consonants linked by a linker */
  uint8_t permitted = unitables_grapheme_break_simple(state_bc, tbc) &&
                      !(state_icb == Unitables_IndicConjunctBreak_Linker &&
                        ticb == Unitables_IndicConjunctBreak_Consonant);

  *state = (uint32_t)unitables_boundclass_next(state_bc, tbc) |
           ((uint32_t)unitables_icb_next(state_icb, ticb) << 8);
  return permitted;
}

File: unitables.py
"""Generate unitables_data.c from the Unicode Character Database.

PROCESS blocks read a UCD file into a table keyed by code point; PRODUCE blocks
combine those into the emitted C arrays. Layout follows utf8proc (see ref/).
"""

import argparse
from collections import namedtuple
from pathlib import Path

parser = argparse.ArgumentParser(
    description="Generate unitables_data.c from the Unicode Character Database."
)
parser.add_argument(
    "output_dir", type=Path, help="directory to write unitables_data.c into"
)
parser.add_argument("--unicode-data", required=True, type=Path)
parser.add_argument("--composition-exclusions", required=True, type=Path)
parser.add_argument("--case-folding", required=True, type=Path)
parser.add_argument("--grapheme-break-property", required=True, type=Path)
parser.add_argument("--emoji-data", required=True, type=Path)
parser.add_argument("--derived-core-properties", required=True, type=Path)
args = parser.parse_args()

MAX_CODEPOINT = 0x110000
PAGE_SIZE = 0x100
SEQ_NONE = 0  # a *_seqindex with no mapping; sequence offset 0 is reserved
COMB_NONE = 0x7FFF  # comb_index value meaning "cannot begin a combining pair"


def intern(store, seen, key, block):
    offset = seen.get(key)
    if offset is None:
        offset = len(store)
        store.extend(block)
        seen[key] = offset
    return offset


# =============================================================================
# PROCESS  UnicodeData.txt
# =============================================================================

UnicodeRecord = namedtuple(
    "UnicodeRecord",
    "category combining_class bidi_class decomp_type decomp "
    "uppercase lowercase titlecase",
)

unicode_data = {}


def decomp_type_enum(tag):
    suffix = "NoBreak" if tag == "noBreak" else tag.capitalize()
    return "Unitables_Decomp_Type_" + suffix


def parse_decomposition(text):
    if not text:
        return "0", None
    parts = text.split()
    if parts[0].startswith("<"):
        return decomp_type_enum(parts[0][1:-1]), [int(p, 16) for p in parts[1:]]
    return "0", [int(p, 16) for p in parts]


def parse_mapping(text):
    return [int(text, 16)] if text else None


def parse_unicode_record(fields):
    decomp_type, decomp = parse_decomposition(fields[5])
    return UnicodeRecord(
        category="Unitables_Category_" + fields[2],
        combining_class=int(fields[3]),
        bidi_class="Unitables_Bidi_Class_" + fields[4],
        decomp_type=decomp_type,
        decomp=decomp,
        uppercase=parse_mapping(fields[12]),
        lowercase=parse_mapping(fields[13]),
        titlecase=parse_mapping(fields[14]),
    )


lines = args.unicode_data.read_text(encoding="utf-8").splitlines()
i = 0
while i < len(lines):
    fields = lines[i].split(";")
    i += 1
    code, name = int(fields[0], 16), fields[1]
    record = parse_unicode_record(fields)
    if name.endswith(", First>"):
        last = int(lines[i].split(";")[0], 16)
        i += 1
        for cp in range(code, last + 1):
            unicode_data[cp] = record
    else:
        unicode_data[code] = record


# Hangul syllables decompose algorithmically; these mirror unitables.c.
HANGUL_SBASE = 0xAC00
HANGUL_LBASE = 0x1100
HANGUL_VBASE = 0x1161
HANGUL_TBASE = 0x11A7
HANGUL_TCOUNT = 28
HANGUL_NCOUNT = 588
HANGUL_SCOUNT = 11172

full_decompositions = {}


def full_decomposition(code, compat):
    """Expand code the way unitables_decompose_into does, recursively."""
    cached = full_decompositions.get((code, compat))
    if cached is not None:
        return cached

    syllable = code - HANGUL_SBASE
    if 0 <= syllable < HANGUL_SCOUNT:
        result = [
            HANGUL_LBASE + syllable // HANGUL_NCOUNT,
            HANGUL_VBASE + (syllable % HANGUL_NCOUNT) // HANGUL_TCOUNT,
        ]
        if syllable % HANGUL_TCOUNT:
            result.append(HANGUL_TBASE + syllable % HANGUL_TCOUNT)
    else:
        rec = unicode_data[code]
        if not rec.decomp or (rec.decomp_type != "0" and not compat):
            result = [code]
        else:
            result = []
            for part in rec.decomp:
                result += full_decomposition(part, compat)

    full_decompositions[(code, compat)] = result
    return result


# Callers size fixed buffers by UNITABLES_DECOMPOSE_MAX, and unitables_decompose
# no longer takes a capacity, so outgrowing it must fail the build.
assert max(len(full_decomposition(c, True)) for c in unicode_data) <= 18, (
    "decomposition exceeds UNITABLES_DECOMPOSE_MAX"
)


# =============================================================================
# PROCESS  CompositionExclusions.txt
# =============================================================================

composition_exclusions = set()
for line in args.composition_exclusions.read_text(encoding="utf-8").splitlines():
    line = line.split("#")[0].strip()
    if line:
        composition_exclusions.add(int(line, 16))


# =============================================================================
# PROCESS  CaseFolding.txt
# =============================================================================


def parse_ucd(path):
    """Parse a UCD semicolon-delimited file. Yields (first, last, fields)
    where fields is a list of stripped strings after the code point range."""
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.split("#")[0].strip()
        if not line:
            continue
        fields = [f.strip() for f in line.split(";")]
        rng = fields[0]
        if ".." in rng:
            first, last = rng.split("..")
            first, last = int(first, 16), int(last, 16)
        else:
            first = last = int(rng, 16)
        yield first, last, fields[1:]


casefold = {}
for first, last, fields in parse_ucd(args.case_folding):
    if fields[0] in ("C", "F"):
        mapping = [int(part, 16) for part in fields[1].split()]
        for cp in range(first, last + 1):
            casefold[cp] = mapping

# Same contract as UNITABLES_DECOMPOSE_MAX above.
assert max(len(v) for v in casefold.values()) <= 3, (
    "case folding exceeds UNITABLES_CASEFOLD_MAX"
)


# =============================================================================
# PROCESS  GraphemeBreakProperty.txt + emoji-data.txt
# =============================================================================

BOUND_CLASS_MAP = {
    "OTHER": "Unitables_Bound_Class_Other",
    "CR": "Unitables_Bound_Class_CR",
    "LF": "Unitables_Bound_Class_LF",
    "CONTROL": "Unitables_Bound_Class_Control",
    "EXTEND": "Unitables_Bound_Class_Extend",
    "L": "Unitables_Bound_Class_L",
    "V": "Unitables_Bound_Class_V",
    "T": "Unitables_Bound_Class_T",
    "LV": "Unitables_Bound_Class_LV",
    "LVT": "Unitables_Bound_Class_LVT",
    "REGIONAL_INDICATOR": "Unitables_Bound_Class_Regional_Indicator",
    "SPACINGMARK": "Unitables_Bound_Class_SpacingMark",
    "PREPEND": "Unitables_Bound_Class_Prepend",
    "ZWJ": "Unitables_Bound_Class_ZWJ",
    "EXTENDED_PICTOGRAPHIC": "Unitables_Bound_Class_Extended_Pictographic",
}

grapheme_bound_class = {}

for first, last, fields in parse_ucd(args.grapheme_break_property):
    key = fields[0].upper().replace(" ", "")
    for cp in range(first, last + 1):
        grapheme_bound_class[cp] = BOUND_CLASS_MAP[key]

for first, last, fields in parse_ucd(args.emoji_data):
    if fields[0] == "Extended_Pictographic":
        for cp in range(first, last + 1):
            grapheme_bound_class[cp] = "Unitables_Bound_Class_Extended_Pictographic"
    elif fields[0] == "Emoji_Modifier":
        for cp in range(first, last + 1):
            grapheme_bound_class[cp] = "Unitables_Bound_Class_Extend"


# =============================================================================
# PROCESS  DerivedCoreProperties.txt  (InCB = Indic_Conjunct_Break)
# =============================================================================

INDIC_CONJUNCT_BREAK_MAP = {
    "Linker": "Unitables_Indic_Conjunct_Break_Linker",
    "Consonant": "Unitables_Indic_Conjunct_Break_Consonant",
    "Extend": "Unitables_Indic_Conjunct_Break_Extend",
}

indic_conjunct_break = {}

for first, last, fields in parse_ucd(args.derived_core_properties):
    if (
        len(fields) >= 2
        and fields[0] == "InCB"
        and fields[1] in INDIC_CONJUNCT_BREAK_MAP
    ):
        for cp in range(first, last + 1):
            indic_conjunct_break[cp] = INDIC_CONJUNCT_BREAK_MAP[fields[1]]


# =============================================================================
# PRODUCE  UNITABLES_COMBINATIONS_SECOND + UNITABLES_COMBINATIONS_COMBINED
# =============================================================================
# Invert the canonical two-code-point decompositions (starter first, not
# excluded) into a table grouped by first code point and sorted by second.

combinations = {}
for code in sorted(unicode_data):
    rec = unicode_data[code]
    if (
        rec.decomp_type == "0"
        and rec.decomp
        and len(rec.decomp) == 2
        and unicode_data[rec.decomp[0]].combining_class == 0
        and code not in composition_exclusions
    ):
        first, second = rec.decomp
        combinations.setdefault(first, {})[second] = code

combinations_second = []
combinations_combined = []
comb_index = {}  # first cp -> start offset in the combination arrays
comb_length = {}  # first cp -> number of entries
comb_issecond = set()
for first in sorted(combinations):
    comb_index[first] = len(combinations_second)
    comb_length[first] = len(combinations[first])
    for second in sorted(combinations[first]):
        combinations_second.append(second)
        combinations_combined.append(combinations[first][second])
        comb_issecond.add(second)

assert len(combinations_second) < COMB_NONE, (
    "combination table exceeds 15-bit comb_index"
)
assert max(comb_length.values()) <= 0xFF, "comb_length exceeds 8 bits"


# =============================================================================
# PRODUCE  UNITABLES_SEQUENCES + UNITABLES_PROPERTIES
# =============================================================================

sequences = [0]  # offset 0 is reserved so no real mapping encodes as SEQ_NONE
sequence_offsets = {}


def encode_sequence(mapping):
    if mapping is None:
        return SEQ_NONE
    units = []
    for cp in mapping:
        if cp <= 0xFFFF:
            units.append(cp)
        else:
            cp -= 0x10000
            units += [0xD800 + (cp >> 10), 0xDC00 + (cp & 0x3FF)]
    length = len(mapping) - 1
    # seqindex = offset (low 14 bits) | length-1 (top 2 bits; 3 = length stored inline).
    block = [length, *units] if length >= 3 else units
    offset = intern(sequences, sequence_offsets, tuple(units), block)
    assert offset <= 0x3FFF, "sequence storage exceeds the 14-bit seqindex range"
    return offset | (min(length, 3) << 14)


SENTINEL = (
    "Unitables_Category_Cn",
    0,
    "0",
    "0",
    SEQ_NONE,
    SEQ_NONE,
    SEQ_NONE,
    SEQ_NONE,
    SEQ_NONE,
    COMB_NONE,
    0,
    0,
    "Unitables_Bound_Class_Other",
    "Unitables_Indic_Conjunct_Break_None",
)
properties = [SENTINEL]
property_indices = {SENTINEL: 0}
char_index = [0] * MAX_CODEPOINT

for code in sorted(unicode_data):
    rec = unicode_data[code]
    entry = (
        rec.category,
        rec.combining_class,
        rec.bidi_class,
        rec.decomp_type,
        encode_sequence(rec.decomp),
        encode_sequence(casefold.get(code)),
        encode_sequence(rec.uppercase),
        encode_sequence(rec.lowercase),
        encode_sequence(rec.titlecase),
        comb_index.get(code, COMB_NONE),
        comb_length.get(code, 0),
        1 if code in comb_issecond else 0,
        grapheme_bound_class.get(code, "Unitables_Bound_Class_Other"),
        indic_conjunct_break.get(code, "Unitables_Indic_Conjunct_Break_None"),
    )
    char_index[code] = intern(properties, property_indices, entry, [entry])


# =============================================================================
# PRODUCE  UNITABLES_STAGE1 + UNITABLES_STAGE2
# =============================================================================
# lookup: stage2[(stage1[cp >> 8] << 8) + (cp & 0xFF)]

stage1 = []
stage2 = []
page_offsets = {}
for start in range(0, MAX_CODEPOINT, PAGE_SIZE):
    page = tuple(char_index[start : start + PAGE_SIZE])
    stage1.append(intern(stage2, page_offsets, page, page) // PAGE_SIZE)

assert max(stage1) <= 0xFFFF, "stage1 page number exceeds uint16"
assert max(stage2) <= 0xFFFF, "property index exceeds uint16"


# =============================================================================
# EMIT  unitables_data.c
# =============================================================================


def format_seq(value):
    return "UNITABLES_SEQ_NONE" if value == SEQ_NONE else str(value)


def c_array(ctype, name, values):
    rows = [", ".join(map(str, values[i : i + 12])) for i in range(0, len(values), 12)]
    return "static {} const {}[] = {{\n  {},\n}};".format(
        ctype, name, ",\n  ".join(rows)
    )


def property_row(entry):
    category, combining_class, bidi_class, decomp_type = entry[:4]
    dseq, cfseq, useq, lseq, tseq = entry[4:9]
    comb_idx, comb_len, comb_2nd = entry[9:12]
    bound_class, icb = entry[12:14]
    comb_idx = "UNITABLES_COMB_NONE" if comb_idx == COMB_NONE else str(comb_idx)
    fields = [
        category,
        str(combining_class),
        bidi_class,
        decomp_type,
        format_seq(dseq),
        format_seq(cfseq),
        format_seq(useq),
        format_seq(lseq),
        format_seq(tseq),
        comb_idx,
        str(comb_len),
        str(comb_2nd),
        bound_class,
        icb,
    ]
    return "  {{ {} }},".format(", ".join(fields))


properties_block = (
    "static struct Unitables_Properties const UNITABLES_PROPERTIES[] = {\n"
    + "\n".join(property_row(e) for e in properties)
    + "\n};"
)

blocks = [
    (
        "/* Generated by unitables.py from UnicodeData.txt, "
        "CompositionExclusions.txt, CaseFolding.txt, GraphemeBreakProperty.txt, "
        "emoji-data.txt, and DerivedCoreProperties.txt. Do not edit. */"
    ),
    c_array("uint16_t", "UNITABLES_SEQUENCES", sequences),
    c_array("uint16_t", "UNITABLES_STAGE1", stage1),
    c_array("uint16_t", "UNITABLES_STAGE2", stage2),
    c_array("int32_t", "UNITABLES_COMBINATIONS_SECOND", combinations_second),
    c_array("int32_t", "UNITABLES_COMBINATIONS_COMBINED", combinations_combined),
    properties_block,
]
(args.output_dir / "unitables_data.c").write_text(
    "\n\n".join(blocks) + "\n", encoding="utf-8"
)
